backend/rest_test.go

```diff
@@ -0,0 +1,172 @@
+package backend_test
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/gorilla/mux"
+	"github.com/restic/restic/backend"
+	"github.com/restic/restic/backend/rest"
+)
+
+func setupRestBackend(t *testing.T) *rest.Rest {
+	url, _ := url.Parse("http://localhost:8000")
+	backend, _ := rest.Open(url)
+	return backend
+}
+
+func TestRestBackend(t *testing.T) {
+	// Initializing a temporary repository for the backend
+	path, _ := ioutil.TempDir("", "restic-repository-")
+	defer os.RemoveAll(path)
+
+	dirs := []string{
+		path,
+		filepath.Join(path, "data"),
+		filepath.Join(path, "snapshot"),
+		filepath.Join(path, "index"),
+		filepath.Join(path, "lock"),
+		filepath.Join(path, "key"),
+		filepath.Join(path, "temp"),
+	}
+
+	for _, d := range dirs {
+		os.MkdirAll(d, backend.Modes.Dir)
+	}
+
+	// Routing the repository requests
+	r := mux.NewRouter()
+
+	// Exists
+	r.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		file := filepath.Join(path, "config")
+		if _, err := os.Stat(file); err != nil {
+			http.Error(w, "No repository here", 404)
+		}
+	}).Methods("HEAD")
+
+	// List blobs
+	r.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		file := filepath.Join(path, "config")
+		if _, err := os.Stat(file); err != nil {
+			http.Error(w, "No repository here", 404)
+		}
+	}).Methods("GET")
+
+	// Head file
+	r.HandleFunc("/{file}", func(w http.ResponseWriter, r *http.Request) {
+		vars := mux.Vars(r)
+		file := filepath.Join(path, vars["file"])
+		if _, err := os.Stat(file); err != nil {
+			http.Error(w, "File not found", 404)
+		}
+	}).Methods("HEAD")
+
+	// Get file
+	r.HandleFunc("/{file}", func(w http.ResponseWriter, r *http.Request) {
+		vars := mux.Vars(r)
+		file := filepath.Join(path, vars["file"])
```

Comment:
Ah, if we're going to use this code, we must carefully check it for path traversals (`../../` etc). This should even be checked for the tests, because while the test is running the server is available, so anybody who can connect to the test REST server could `GET` or `DELETE` arbitrary files from the system. That's not good :)

Cleanup might look something like this: https://github.com/golang/tools/blob/master/godoc/server.go#L238-L239